src/toml.rs

```diff
@@ -0,0 +1,228 @@
+//! Support for variable substitution in TOML data.
+
+use serde::de::DeserializeOwned;
+
+use crate::VariableMap;
+
+/// Parse a struct from TOML data, after performing variable substitution on string values.
+///
+/// This function first parses the data into a [`toml::Value`],
+/// then performs variable substitution on all string values,
+/// and then parses it further into the desired type.
+pub fn from_slice<'a, T: DeserializeOwned, M>(data: &[u8], variables: &'a M) -> Result<T, Error>
+where
+	M: VariableMap<'a> + ?Sized,
+	M::Value: AsRef<str>,
+{
+	from_str(std::str::from_utf8(data)?, variables)
+}
+
+/// Parse a struct from TOML data, after performing variable substitution on string values.
+///
+/// This function first parses the data into a [`toml::Value`],
+/// then performs variable substitution on all string values,
+/// and then parses it further into the desired type.
+pub fn from_str<'a, T: DeserializeOwned, M>(data: &str, variables: &'a M) -> Result<T, Error>
+where
+	M: VariableMap<'a> + ?Sized,
+	M::Value: AsRef<str>,
+{
+	let mut value: toml::Value = toml::from_str(data)?;
```

Comment:
Why does it make sense to first interpret the string as a `toml::Value`, instead of doing the substitution directly on the `str`? Assuming there is a reason, could you add it as a comment?

Reply:
The result of that is badly defined. What if the substitution result messes up the TOML syntax?

Reply:
It would be fine to return an error in that case. 

In addition, I'm not sure if you should, but with the current implementation this is not possible:

```toml
[$foo]
bar = "baz"
```

I also have no clue if this is possible in the yaml implementation to be fair.

Reply:
> It would be fine to return an error in that case.

No, it should just perform substitution on strings. Messing up the string before parsing is horrible.



> In addition, I'm not sure if you should, but with the current implementation this is not possible:
> 
> ```toml
> [$foo]
> bar = "baz"
> ```

It's not, and this is documented. The substitution is performed on string values in the YAML/TOML document.

Reply:
> No, it should just perform substitution on strings. Messing up the string before parsing is horrible.

I'm not sure what your argument is, besides that it's horrible. Why is it horrible?

> It's not, and this is documented. The substitution is performed on string values in the YAML/TOML document.

In that case I'm fine with keeping the same features in TOML vs YAML.

Reply:
> I'm not sure what your argument is, besides that it's horrible. Why is it horrible?

Because it will break parsing depending on the value of your substitution parameters. You will suddenly need to employ TOML escaping in your variables.

And if an application uses the environment to allow users to configure certain variables, it's even a security risk because it will allow users to potentially alter the structure of the TOML file and configure fields that never even contained a placeholder.

Reply:
Okay yeah fair, hadn't considered "toml injection". I would not have been opposed to an additional function that replaces variables across the entire file, as long as the risk is clear, but let's not add that.

In that case, LGTM :+1: